**Import data**

In [2]:
import numpy as np
import scipy as sp
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

# Pandas options
pd.set_option('display.max_colwidth', 1000,'display.max_rows',None, 'display.max_columns',None)

# Plot options
%matplotlib inline
mpl.style.use('ggplot')
sns.set(style = 'whitegrid')


In [7]:
import os
os.listdir(os.getcwd())

['LendingLoanEDA.ipynb',
 'README.md',
 '.ipynb_checkpoints',
 '.git',
 'accepted_2007_to_2018Q4.csv.gz']

In [17]:
loans = pd.read_csv('accepted_2007_to_2018Q4.csv.gz', compression ='gzip',low_memory=True)

/var/folders/8f/c3wp81s97mx_4cyjt0ks4rc80000gn/T/ipykernel_46483/393940706.py:1: DtypeWarning: Columns (0,19,49,59,118,129,130,131,134,135,136,139,145,146,147) have mixed types. Specify dtype option on import or set low_memory=False.
  loans = pd.read_csv('accepted_2007_to_2018Q4.csv.gz', compression ='gzip',low_memory=True)


Check basic dataframe info:

In [8]:
loans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2260701 entries, 0 to 2260700
Columns: 151 entries, id to settlement_term
dtypes: float64(113), object(38)
memory usage: 2.5+ GB


There are 2.3 million rows and 151 variables. The size of the dataset is 2.5+GB

In [9]:
loans.sample(5)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
598036   114023483        NaN    12000.0      12000.0          12000.0   
1805710    6909113        NaN    25975.0      25975.0          25975.0   
310824    46367660        NaN     3000.0       3000.0           3000.0   
2180925   94225554        NaN    13200.0      13200.0          13200.0   
1086551   70091839        NaN    18000.0      18000.0          18000.0   

               term  int_rate  installment grade sub_grade  \
598036    36 months     13.59       407.75     C        C2   
1805710   60 months     25.28       766.68     F        F5   
310824    36 months      9.17        95.64     B        B2   
2180925   60 months     17.99       335.13     D        D2   
1086551   36 months     10.78       587.43     B        B4   

                            emp_title emp_length home_ownership  annual_inc  \
598036             Clinical Director     3 years            OWN    125000.0   
1805710                     U.S. ARMY  10+ years       MORTGAGE     58000.0   
310824   Electro Mechanical Assembler  10+ years       MORTGAGE     35000.0   
2180925                    Estimating  10+ years       MORTGAGE     78000.0   
1086551        Clinical Nurse Manager    2 years           RENT    113000.0   

        verification_status   issue_d loan_status pymnt_plan  \
598036      Source Verified  Aug-2017     Current          n   
1805710            Verified  Sep-2013  Fully Paid          n   
310824         Not Verified  Apr-2015  Fully Paid          n   
2180925     Source Verified  Dec-2016  Fully Paid          n   
1086551     Source Verified  Jan-2016  Fully Paid          n   

                                                                        url  \
598036   https://lendingclub.com/browse/loanDetail.action?loan_id=114023483   
1805710    https://lendingclub.com/browse/loanDetail.action?loan_id=6909113   
310824    https://lendingclub.com/browse/loanDetail.action?loan_id=46367660   
2180925   https://lendingclub.com/browse/loanDetail.action?loan_id=94225554   
1086551   https://lendingclub.com/browse/loanDetail.action?loan_id=70091839   

        desc             purpose                    title zip_code addr_state  \
598036   NaN    home_improvement         Home improvement    373xx         TN   
1805710  NaN  debt_consolidation       Debt consolidation    765xx         TX   
310824   NaN  debt_consolidation       Debt consolidation    841xx         UT   
2180925  NaN  debt_consolidation       Debt consolidation    460xx         IN   
1086551  NaN         credit_card  Credit card refinancing    928xx         CA   

           dti  delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
598036    7.81          3.0         Nov-2001           680.0            684.0   
1805710  22.54          0.0         Mar-2000           670.0            674.0   
310824   33.02          0.0         Oct-1991           730.0            734.0   
2180925  17.74          0.0         Apr-1981           690.0            694.0   
1086551   4.41          0.0         Aug-2000           670.0            674.0   

         inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
598036              2.0                     7.0                     NaN   
1805710             4.0                     NaN                     NaN   
310824              0.0                     NaN                     NaN   
2180925             0.0                     NaN                    79.0   
1086551             1.0                     NaN                     NaN   

         open_acc  pub_rec  revol_bal  revol_util  total_acc  \
598036       11.0      0.0      807.0        12.0       31.0   
1805710      10.0      0.0    10206.0        49.3       31.0   
310824        7.0      0.0    21156.0        66.7       14.0   
2180925       9.0      1.0     8576.0        42.9       22.0   
1086551      13.0      0.0    18519.0        82.3       37.0   

        initial_list_status  out_prncp  out_prncp_inv

**Response variable: loan status**

In [10]:
loans['loan_status'].value_counts(dropna=False)

Fully Paid                                             1076751
Current                                                 878317
Charged Off                                             268559
Late (31-120 days)                                       21467
In Grace Period                                           8436
Late (16-30 days)                                         4349
Does not meet the credit policy. Status:Fully Paid        1988
Does not meet the credit policy. Status:Charged Off        761
Default                                                     40
NaN                                                         33
Name: loan_status, dtype: int64

For the simplicity of this project, we only consider the loan that is fully paid or charged off.

In [18]:
loans = loans.loc[loans['loan_status'].isin(['Fully Paid','Charged Off'])]

In [19]:
loans.shape # There are 1.3 million loans remaining

(1345310, 151)

In [20]:
loans['loan_status'].value_counts()

Fully Paid     1076751
Charged Off     268559
Name: loan_status, dtype: int64

In [21]:
# status counts as percentage
loans['loan_status'].value_counts(normalize=True)

Fully Paid     0.800374
Charged Off    0.199626
Name: loan_status, dtype: float64

About 80% of the remaining loans have been fully paid and 20% have charged off, so we have a somewhat unbalanced classification problem.

**Feature selection**

The full dataset has 151 features for each loan. We will select features in two steps:
1. Drop features with more than 30% of their data missing
2. Of the remaining features, choose only those that would be available to an investor before deciding to fund the loan.

1. Drop the features missing more than 30% data

In [22]:
# Calculate the percentage of missing data for each feature
missing_fraction = loans.isnull().mean().sort_values(ascending=False)
# Top 10 features missing the most data:
missing_fraction.head(10)

next_pymnt_d                                  1.000000
member_id                                     1.000000
orig_projected_additional_accrued_interest    0.997206
hardship_start_date                           0.995723
hardship_length                               0.995723
hardship_type                                 0.995723
hardship_reason                               0.995723
hardship_status                               0.995723
deferral_term                                 0.995723
hardship_amount                               0.995723
dtype: float64

In [32]:
# store all variables missing more than 30% data:
drop_list = sorted(list(missing_fraction[missing_fraction > 0.3].index))
len(drop_list) # 58 features will be dropped
loans.drop(labels=drop_list, axis = 1, inplace = True)
loans.shape # (1345310, 93)


(1345310, 93)

2. Only keep loan features known to potential investors

In [33]:
print(sorted(loans.columns))

['acc_now_delinq', 'acc_open_past_24mths', 'addr_state', 'annual_inc', 'application_type', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 'chargeoff_within_12_mths', 'collection_recovery_fee', 'collections_12_mths_ex_med', 'debt_settlement_flag', 'delinq_2yrs', 'delinq_amnt', 'disbursement_method', 'dti', 'earliest_cr_line', 'emp_length', 'emp_title', 'fico_range_high', 'fico_range_low', 'funded_amnt', 'funded_amnt_inv', 'grade', 'hardship_flag', 'home_ownership', 'id', 'initial_list_status', 'inq_last_6mths', 'installment', 'int_rate', 'issue_d', 'last_credit_pull_d', 'last_fico_range_high', 'last_fico_range_low', 'last_pymnt_amnt', 'last_pymnt_d', 'loan_amnt', 'loan_status', 'mo_sin_old_il_acct', 'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl', 'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_inq', 'num_accts_ever_120_pd', 'num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats', 'num_bc_tl', 'num_il_tl', 'num_op_rev_tl', 'num_rev_accts', 'num_rev_tl_bal_gt_0', 'num_sa